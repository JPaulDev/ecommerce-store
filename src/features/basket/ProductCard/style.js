import styled, { css } from 'styled-components';

const overflowEllipsisMixin = css`
  -webkit-box-orient: vertical;
  display: -webkit-box;
  overflow: hidden;
`;

export const buttonMixin = css`
  background-image: linear-gradient(
    to bottom,
    ${({ theme }) => theme.colors.green[500]},
    ${({ theme }) => theme.colors.green[700]}
  );
  border-radius: ${({ theme }) => theme.borderRadius.lg};
  box-shadow: ${({ theme }) => theme.boxShadows.md};
  color: white;
  cursor: pointer;
  font-weight: 700;
  margin-top: 10px;
  padding: 13px;
  font-size: ${({ theme }) => theme.fontSizes[14]};
  text-transform: uppercase;

  &:disabled {
    background: ${({ theme }) => theme.colors.neutral[300]};
    cursor: not-allowed;
    box-shadow: none;
  }

  &:not(:disabled):hover {
    box-shadow: inset 0 0 0 100px rgb(0 0 0 / 10%),
      ${({ theme }) => theme.boxShadows.hover};
  }

  &:not(:disabled):active {
    transform: scale(0.96);
  }
`;

export const HorizontalCard = styled.li`
  --border-color: ${({ theme }) => theme.colors.stone[400]};

  border: 1px solid var(--border-color);
  display: grid;
  grid-template-columns: auto 1fr;
  padding: 20px 25px;

  > img {
    height: auto;
    width: 120px;
  }

  @media ${({ theme }) => theme.breakpoints.mobileLg} {
    > img {
      width: 150px;
    }
  }

  @media ${({ theme }) => theme.breakpoints.tablet} {
    grid-template-columns: auto 1fr 280px;
  }

  @media ${({ theme }) => theme.breakpoints.desktopXs} {
    > img {
      width: 180px;
    }
  }
`;

export const LeftContainer = styled.div`
  padding: 10px 0 10px 25px;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  row-gap: 10px;

  > img {
    height: auto;
    width: 70px;
  }

  .sale-tag {
    background-color: ${({ theme }) => theme.colors.red[500]};
    border-radius: ${({ theme }) => theme.borderRadius.lg};
    color: white;
    font-size: ${({ theme }) => theme.fontSizes[14]};
    font-weight: 600;
    padding: 3px 10px;
  }

  > p {
    color: ${({ theme }) => theme.colors.stone[600]};
    font-size: ${({ theme }) => theme.fontSizes[15]};
    line-height: 1.5;
  }

  @media ${({ theme }) => theme.breakpoints.tablet} {
    padding: 10px 25px;
  }

  @media ${({ theme }) => theme.breakpoints.desktopXs} {
    > img {
      width: 90px;
    }
  }
`;

export const RightContainer = styled.div`
  border-top: 2px solid var(--border-color);
  grid-column: 1 / 3;
  padding-top: 10px;
  display: flex;
  justify-content: space-between;
  margin-top: 10px;

  .price-container {
    display: flex;
    flex-direction: column;
    position: relative;
  }

  button {
    ${buttonMixin}
    min-width: 160px;
  }

  .container {
    align-items: flex-end;
    flex-direction: column;
    display: flex;
    justify-content: space-between;
    margin-top: 10px;
    row-gap: 5px;
  }

  @media ${({ theme }) => theme.breakpoints.tabletSm} {
    margin-left: 25px;
    grid-column: 2 / 3;
  }

  @media ${({ theme }) => theme.breakpoints.tablet} {
    padding-left: 25px;
    padding-top: 0;
    margin: 0;
    border-left: 1px solid var(--border-color);
    flex-direction: column;
    justify-content: space-evenly;
    grid-column: auto;
    border-top: none;

    .quantity-wrapper {
      position: absolute;
      right: 0;
      top: 0;
    }

    button {
      width: 100%;
    }

    .container {
      flex-direction: row;
      align-items: center;
    }
  }
`;

export const VerticalCard = styled.li`
  align-items: center;
  display: flex;
  justify-content: center;
  background-color: white;
  border: 1px solid ${({ theme }) => theme.colors.stone[200]};
  border-radius: ${({ theme }) => theme.borderRadius.xl};
  flex-direction: column;
  padding: 20px;
  text-align: center;

  > img {
    height: auto;
    width: 150px;
  }

  .product-name {
    ${overflowEllipsisMixin}
    -webkit-line-clamp: 2;
  }

  .product-description {
    ${overflowEllipsisMixin}
    color: ${({ theme }) => theme.colors.stone[600]};
    font-size: ${({ theme }) => theme.fontSizes[14]};
    -webkit-line-clamp: 3;
    line-height: 1.5;
    margin: 10px 0;
  }

  .divider {
    height: 2px;
    background-color: ${({ theme }) => theme.colors.stone[300]};
    margin: 10px 0;
    width: 100%;
  }

  > button {
    ${buttonMixin}
    min-width: 180px;
  }

  @media ${({ theme }) => theme.breakpoints.mobileLg} {
    min-height: 520px;
  }
`;
