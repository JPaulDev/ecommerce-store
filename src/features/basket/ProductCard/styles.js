import styled, { css } from 'styled-components';

const flexMixin = css`
  align-items: center;
  display: flex;
  justify-content: center;
`;

const overflowEllipsisMixin = css`
  -webkit-box-orient: vertical;
  display: -webkit-box;
  overflow: hidden;
  text-align: center;
`;

const imageMixin = css`
  height: auto;
  max-width: 100%;
`;

const containerMixin = css`
  display: flex;
  justify-content: space-between;
  margin-top: 10px;
`;

export const ListItem = styled.li`
  ${({ variant }) => {
    if (variant === 'horizontal') {
      return css`
        border: 1px solid var(--color-grey-9);
        display: grid;
        grid-template-columns: auto 1fr;
        margin-bottom: 15px;
        padding: 20px 25px;

        @media screen and (min-width: 800px) {
          grid-template-columns: auto 1fr 280px;
        }
      `;
    }

    return css`
      background-color: white;
      border: 1px solid var(--color-grey-12);
      border-radius: 10px;
      flex-direction: column;
      height: 100%;
      padding: 20px;
      ${flexMixin}

      img {
        height: auto;
        width: 150px;
      }

      :nth-child(4n) {
        display: none;

        @media screen and (min-width: 1110px) {
          display: flex;
        }
      }

      @media screen and (min-width: 570px) {
        min-height: 510px;
      }
    `;
  }}
`;

export const ProductName = styled.p`
  -webkit-line-clamp: 2;
  margin: 5px 0 10px;
  ${overflowEllipsisMixin}

  @media screen and (min-width: 570px) {
    margin: 5px 0 auto;
  }
`;

export const ProductDescription = styled.p`
  color: var(--color-grey-4);
  font-size: var(--font-size-14);

  ${({ variant }) => {
    if (variant === 'horizontal') {
      return css`
        margin: 5px 0;
      `;
    }

    return css`
      -webkit-line-clamp: 3;
      line-height: 1.5;
      margin-top: 10px;
      ${overflowEllipsisMixin}
    `;
  }}
`;

export const PartNumberWrapper = styled.div`
  border-bottom: 2px solid var(--color-grey-11);
  margin: 5px 0;
  padding-bottom: 10px;
  width: 100%;
  ${flexMixin}
`;

export const ImageWrapper = styled.div`
  height: auto;
  width: 120px;

  img {
    ${imageMixin}
  }

  @media screen and (min-width: 530px) {
    width: 150px;
    height: auto;
  }

  @media screen and (min-width: 930px) {
    width: 180px;
    height: auto;
  }
`;

export const LeftContainer = styled.div`
  padding: 10px 0 10px 25px;

  @media screen and (min-width: 800px) {
    padding: 10px 25px;
  }
`;

export const LogoWrapper = styled.div`
  height: auto;
  width: 75px;

  img {
    ${imageMixin}
  }

  @media screen and (min-width: 930px) {
    width: 90px;
    height: auto;
  }
`;

export const RightContainer = styled.div`
  border-top: 2px solid var(--color-grey-9);
  grid-column: 1/3;
  padding-top: 10px;
  ${containerMixin}

  @media screen and (min-width: 650px) {
    margin-left: 25px;
    grid-column: 2/3;
  }

  @media screen and (min-width: 800px) {
    padding-left: 25px;
    padding-top: 0;
    margin: 0;
    border-left: 1px solid var(--color-grey-9);
    flex-direction: column;
    justify-content: space-evenly;
    grid-column: initial;
    border-top: none;
  }
`;

export const PriceContainer = styled.div`
  display: flex;
  flex-direction: column;
  position: relative;
`;

export const QuantityWrapper = styled.div`
  @media screen and (min-width: 800px) {
    position: absolute;
    right: 0;
    top: 0;
  }
`;

export const ButtonContainer = styled.div`
  width: 160px;

  @media screen and (min-width: 650px) {
    width: 200px;
  }

  @media screen and (min-width: 800px) {
    width: 100%;
  }
`;

export const InnerContainer = styled.div`
  align-items: flex-end;
  flex-direction: column;
  ${containerMixin}

  @media screen and (min-width: 800px) {
    flex-direction: row;
    align-items: center;
  }
`;

export const SaleTag = styled.div`
  background-color: var(--color-red-3);
  border-radius: 5px;
  color: white;
  display: inline-block;
  font-size: var(--font-size-14);
  font-weight: var(--font-weight-semibold);
  margin-top: 5px;
  padding: 2px 10px;
  text-transform: uppercase;
`;
