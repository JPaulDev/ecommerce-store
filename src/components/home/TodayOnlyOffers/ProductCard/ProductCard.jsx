import Image from 'next/future/image';
import { useState, forwardRef } from 'react';
import { useDispatch } from 'react-redux';
import { addProduct } from '../../../../features/basket/basketSlice';
import usePrice from '../../../../hooks/usePrice';
import {
  Price,
  PreviousPrice,
  PartNumber,
  QuantityDropdown,
  StockIndicator,
  AddToBasket,
} from '../../../common';
import * as Styled from './styles';

function ProductCard({ product = {} }, ref) {
  const [quantity, setQuantity] = useState(1);
  const { price, salePrice } = usePrice(product);
  const dispatch = useDispatch();
  const isOutOfStock = product.stock <= 0;

  const handleChangeQuantity = (e) => setQuantity(parseInt(e.target.value, 10));
  const handleAddProduct = () => {
    dispatch(addProduct(product, quantity));
    setQuantity(1);
  };

  return (
    <Styled.Wrapper ref={ref} data-testid="product">
      <Styled.Container>
        <Styled.ImageWrapper>
          <Image src={product.image} alt="" quality={85} />
        </Styled.ImageWrapper>
        <Styled.ProductName>{product.name}</Styled.ProductName>
        <Styled.ProductDescription>
          {product.description?.join(', ')}
        </Styled.ProductDescription>
        <Styled.PartNumberWrapper>
          <PartNumber fontSize="0.75rem" sku={product.sku} />
        </Styled.PartNumberWrapper>
        <Price price={salePrice || price} fontSize="1.9rem" />
        {salePrice && <PreviousPrice price={price} fontSize="0.93rem" />}
        <QuantityDropdown
          quantity={quantity}
          isDisabled={isOutOfStock}
          onChangeQuantity={handleChangeQuantity}
        />
        <StockIndicator stock={product.stock} marginTop="8px" />
        <AddToBasket
          isDisabled={isOutOfStock}
          width="160px"
          fontSize="0.75rem"
          onAddProduct={handleAddProduct}
        />
      </Styled.Container>
    </Styled.Wrapper>
  );
}

export default forwardRef(ProductCard);
