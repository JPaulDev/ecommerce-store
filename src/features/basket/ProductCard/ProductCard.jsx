import dynamic from 'next/dynamic';
import Image from 'next/future/image';
import { useState, forwardRef } from 'react';
import { useDispatch } from 'react-redux';
import { addProduct } from '../basketSlice';
import { usePrice } from '../../../hooks';
import {
  Price,
  PreviousPrice,
  PartNumber,
  Quantity,
  StockDisplay,
  AddToBasket,
} from '../../../components/product';
import * as Styled from './styles';

const DeliveryDate = dynamic(
  () => import('../../../components/product/DeliveryDate'),
  {
    ssr: false,
  }
);

function ProductCard({ product, variant = 'horizontal' }, ref) {
  const [quantity, setQuantity] = useState(1);
  const dispatch = useDispatch();

  const { price, previousPrice } = usePrice({
    isOnSale: product.isOnSale,
    discount: product.discount,
    price: product.price,
    currencyCode: product.currencyCode,
  });

  const handleChangeQuantity = (e) => setQuantity(Number(e.target.value));

  const handleAddProduct = () => {
    dispatch(addProduct(product, quantity));
    setQuantity(1);
  };

  return (
    <>
      {variant === 'horizontal' && (
        <Styled.ListItem ref={ref} variant={variant}>
          <Styled.ImageWrapper>
            <Image
              src={product.imageUrl}
              alt=""
              width={200}
              height={200}
              quality={90}
              priority="true"
            />
          </Styled.ImageWrapper>
          <Styled.LeftContainer>
            <Styled.LogoWrapper>
              <Image
                src={product.brand}
                alt=""
                width={110}
                height={24}
                quality={90}
              />
            </Styled.LogoWrapper>
            {product.isOnSale && <Styled.SaleTag>SALE</Styled.SaleTag>}
            <Styled.ProductDescription variant={variant}>
              {product.name}, {product.description}
            </Styled.ProductDescription>
            <PartNumber fontSize={13}>{product.sku}</PartNumber>
          </Styled.LeftContainer>
          <Styled.RightContainer>
            <Styled.PriceContainer>
              <Price fontSize={32}>{price}</Price>
              {product.isOnSale && (
                <PreviousPrice>{previousPrice}</PreviousPrice>
              )}
              <Styled.QuantityWrapper>
                <Quantity
                  quantity={quantity}
                  onChangeQuantity={handleChangeQuantity}
                />
              </Styled.QuantityWrapper>
            </Styled.PriceContainer>
            <Styled.ButtonContainer>
              <AddToBasket
                type="button"
                width="100%"
                fontSize={14}
                onAddProduct={handleAddProduct}
              />
              <Styled.InnerContainer>
                <StockDisplay stockStatus={product.stockStatus} />
                <DeliveryDate stockStatus={product.stockStatus} />
              </Styled.InnerContainer>
            </Styled.ButtonContainer>
          </Styled.RightContainer>
        </Styled.ListItem>
      )}

      {variant === 'vertical' && (
        <Styled.ListItem ref={ref} variant={variant}>
          <Image
            src={product.imageUrl}
            alt=""
            width={200}
            height={200}
            quality={90}
          />
          <Styled.ProductName>{product.name}</Styled.ProductName>
          <Styled.ProductDescription variant={variant}>
            {product.description}
          </Styled.ProductDescription>
          <Styled.PartNumberWrapper>
            <PartNumber fontSize={12}>{product.sku}</PartNumber>
          </Styled.PartNumberWrapper>
          <Price fontSize={30}>{price}</Price>
          {product.isOnSale && <PreviousPrice>{previousPrice}</PreviousPrice>}
          <Quantity
            quantity={quantity}
            isDisabled={product.stockStatus === 0}
            onChangeQuantity={handleChangeQuantity}
          />
          <StockDisplay stockStatus={product.stockStatus} marginTop="8px" />
          <AddToBasket
            isDisabled={product.stockStatus === 0}
            width="180px"
            fontSize={12}
            onAddProduct={handleAddProduct}
          />
        </Styled.ListItem>
      )}
    </>
  );
}

export default forwardRef(ProductCard);
