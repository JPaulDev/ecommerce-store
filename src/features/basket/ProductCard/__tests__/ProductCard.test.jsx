import userEvent from '@testing-library/user-event';
import { render, screen } from '../../../../../test-utils';
import ProductCard from '../index';

// Prevents errors caused by next image component.
jest.mock('next/future/image', () => 'img');

const productData = {
  sku: 'GPU-3PAQ-ASU',
  name: 'ASUS ROG Strix GeForce RTX 3080 V2 OC',
  imageUrl: '',
  price: 99.99,
  discount: 30,
  description: '1440MHz, 1905MHz Boost, 8704 CUDA Cores, 10GB GDDR6X',
  stockStatus: 2,
  brand: '',
  currencyCode: 'GBP',
  isOnSale: true,
};

function setup(component, preloadedState) {
  return {
    user: userEvent.setup(),
    ...render(component, preloadedState),
  };
}

describe('ProductCard vertical variant', () => {
  it('should render the vertical variant', () => {
    setup(<ProductCard product={productData} variant="vertical" />);

    expect(screen.queryByTestId('horizontal-card')).not.toBeInTheDocument();
    expect(screen.getByTestId('vertical-card')).toBeInTheDocument();
  });

  describe('quantity input', () => {
    it('should record user changes to quantity', async () => {
      const { user } = setup(
        <ProductCard product={productData} variant="vertical" />
      );
      const select = screen.getByLabelText(/quantity/i);

      expect(select).toHaveValue('1');

      await user.selectOptions(select, '3');

      expect(select).toHaveValue('3');
    });

    it('should be enabled when the product is in stock', () => {
      setup(<ProductCard product={productData} variant="vertical" />);

      expect(screen.getByLabelText(/quantity/i)).toBeEnabled();
    });

    it('should be enabled when the product is low stock', () => {
      setup(
        <ProductCard
          product={{ ...productData, stockStatus: 1 }}
          variant="vertical"
        />
      );

      expect(screen.getByLabelText(/quantity/i)).toBeEnabled();
    });

    it('should be disabled when the product is out of stock', () => {
      setup(
        <ProductCard
          product={{ ...productData, stockStatus: 0 }}
          variant="vertical"
        />
      );

      expect(screen.getByLabelText(/quantity/i)).toBeDisabled();
    });

    it('should reset back to 1 after clicking add to basket', async () => {
      const { user } = setup(
        <ProductCard product={productData} variant="vertical" />
      );
      const select = screen.getByLabelText(/quantity/i);

      await user.selectOptions(select, '3');

      expect(select).toHaveValue('3');

      await user.click(screen.getByRole('button', { name: /add to basket/i }));

      expect(select).toHaveValue('1');
    });
  });

  describe('add to basket button', () => {
    it('should be enabled when the product is in stock', () => {
      setup(<ProductCard product={productData} variant="vertical" />);

      expect(
        screen.getByRole('button', { name: /add to basket/i })
      ).toBeEnabled();
    });

    it('should be enabled when the product is low stock', () => {
      setup(
        <ProductCard
          product={{ ...productData, stockStatus: 1 }}
          variant="vertical"
        />
      );

      expect(
        screen.getByRole('button', { name: /add to basket/i })
      ).toBeEnabled();
    });

    it('should be disabled when the product is out of stock', () => {
      setup(
        <ProductCard
          product={{ ...productData, stockStatus: 0 }}
          variant="vertical"
        />
      );

      expect(
        screen.getByRole('button', { name: /add to basket/i })
      ).toBeDisabled();
    });

    it('should add 1 of the product to the basket when clicked', async () => {
      const { user, store } = setup(
        <ProductCard product={productData} variant="vertical" />
      );

      expect(store.getState().basket).toMatchObject([]);

      await user.click(screen.getByRole('button', { name: /add to basket/i }));

      expect(store.getState().basket).toMatchObject([
        { ...productData, quantity: 1 },
      ]);
    });

    it('should add the correct quantity to the basket when clicked', async () => {
      const { user, store } = setup(
        <ProductCard product={productData} variant="vertical" />
      );

      expect(store.getState().basket).toMatchObject([]);

      await user.selectOptions(screen.getByLabelText(/quantity/i), '3');
      await user.click(screen.getByRole('button', { name: /add to basket/i }));

      expect(store.getState().basket).toMatchObject([
        { ...productData, quantity: 3 },
      ]);
    });
  });

  describe('stock indicator', () => {
    it('should display in stock', () => {
      setup(<ProductCard product={productData} variant="vertical" />);

      expect(screen.getByText(/in stock/i)).toBeInTheDocument();
    });

    it('should display low stock', () => {
      setup(
        <ProductCard
          product={{ ...productData, stockStatus: 1 }}
          variant="vertical"
        />
      );

      expect(screen.getByText(/low stock/i)).toBeInTheDocument();
    });

    it('should display out of stock', () => {
      setup(
        <ProductCard
          product={{ ...productData, stockStatus: 0 }}
          variant="vertical"
        />
      );

      expect(screen.getByText(/out of stock/i)).toBeInTheDocument();
    });
  });

  describe('price display', () => {
    it('should only display the base price if not on sale', () => {
      setup(
        <ProductCard
          product={{ ...productData, isOnSale: false }}
          variant="vertical"
        />
      );

      expect(screen.getByText('99')).toHaveTextContent('Â£99.99');
      expect(screen.queryByText(/was:/i)).not.toBeInTheDocument();
    });

    it('should display the price and previous price when on sale', () => {
      setup(<ProductCard product={productData} variant="vertical" />);

      expect(screen.getByText('99')).toHaveTextContent('Â£99.99');
      expect(screen.getByText('Â£129.99')).toBeInTheDocument();
    });
  });
});

describe('ProductCard horizontal variant', () => {
  it('should render the horizontal variant', () => {
    setup(<ProductCard product={productData} variant="horizontal" />);

    expect(screen.queryByTestId('vertical-card')).not.toBeInTheDocument();
    expect(screen.getByTestId('horizontal-card')).toBeInTheDocument();
  });

  describe('quantity input', () => {
    it('should record user changes to quantity', async () => {
      const { user } = setup(
        <ProductCard product={productData} variant="horizontal" />
      );
      const select = screen.getByLabelText(/quantity/i);

      expect(select).toHaveValue('1');

      await user.selectOptions(select, '3');

      expect(select).toHaveValue('3');
    });

    it('should be enabled when the product is in stock', () => {
      setup(<ProductCard product={productData} variant="horizontal" />);

      expect(screen.getByLabelText(/quantity/i)).toBeEnabled();
    });

    it('should be enabled when the product is low stock', () => {
      setup(
        <ProductCard
          product={{ ...productData, stockStatus: 1 }}
          variant="horizontal"
        />
      );

      expect(screen.getByLabelText(/quantity/i)).toBeEnabled();
    });

    it('should be enabled when the product is out of stock', () => {
      setup(
        <ProductCard
          product={{ ...productData, stockStatus: 0 }}
          variant="horizontal"
        />
      );

      expect(screen.getByLabelText(/quantity/i)).toBeEnabled();
    });

    it('should reset back to 1 after clicking add to basket', async () => {
      const { user } = setup(
        <ProductCard product={productData} variant="horizontal" />
      );
      const select = screen.getByLabelText(/quantity/i);

      await user.selectOptions(select, '3');

      expect(select).not.toHaveValue('1');

      await user.click(screen.getByRole('button', { name: /add to basket/i }));

      expect(select).toHaveValue('1');
    });
  });

  describe('add to basket button', () => {
    it('should be enabled when the product is in stock', () => {
      setup(<ProductCard product={productData} variant="horizontal" />);

      expect(
        screen.getByRole('button', { name: /add to basket/i })
      ).toBeEnabled();
    });

    it('should be enabled when the product is low stock', () => {
      setup(
        <ProductCard
          product={{ ...productData, stockStatus: 1 }}
          variant="horizontal"
        />
      );

      expect(
        screen.getByRole('button', { name: /add to basket/i })
      ).toBeEnabled();
    });

    it('should be enabled when the product is out of stock', () => {
      setup(
        <ProductCard
          product={{ ...productData, stockStatus: 0 }}
          variant="horizontal"
        />
      );

      expect(
        screen.getByRole('button', { name: /add to basket/i })
      ).toBeEnabled();
    });

    it('should add 1 of the product to the basket when clicked', async () => {
      const { user, store } = setup(
        <ProductCard product={productData} variant="horizontal" />
      );

      expect(store.getState().basket).toMatchObject([]);

      await user.click(screen.getByRole('button', { name: /add to basket/i }));

      expect(store.getState().basket).toMatchObject([
        { ...productData, quantity: 1 },
      ]);
    });

    it('should add the correct quantity to the basket when clicked', async () => {
      const { user, store } = setup(
        <ProductCard product={productData} variant="horizontal" />
      );

      expect(store.getState().basket).toMatchObject([]);

      await user.selectOptions(screen.getByLabelText(/quantity/i), '3');
      await user.click(screen.getByRole('button', { name: /add to basket/i }));

      expect(store.getState().basket).toMatchObject([
        { ...productData, quantity: 3 },
      ]);
    });
  });

  describe('stock indicator', () => {
    it('should display in stock', () => {
      setup(<ProductCard product={productData} variant="horizontal" />);

      expect(screen.getByText(/in stock/i)).toBeInTheDocument();
    });

    it('should display low stock', () => {
      setup(
        <ProductCard
          product={{ ...productData, stockStatus: 1 }}
          variant="horizontal"
        />
      );

      expect(screen.getByText(/low stock/i)).toBeInTheDocument();
    });

    it('should display out of stock', () => {
      setup(
        <ProductCard
          product={{ ...productData, stockStatus: 0 }}
          variant="horizontal"
        />
      );

      expect(screen.getByText(/out of stock/i)).toBeInTheDocument();
    });
  });

  describe('price display', () => {
    it('should only display the base price if not on sale', () => {
      setup(
        <ProductCard
          product={{ ...productData, isOnSale: false }}
          variant="horizontal"
        />
      );

      expect(screen.getByText('99')).toHaveTextContent('Â£99.99');
      expect(screen.queryByText(/was:/i)).not.toBeInTheDocument();
    });

    it('should display the price and previous price when on sale', () => {
      setup(<ProductCard product={productData} variant="horizontal" />);

      expect(screen.getByText('99')).toHaveTextContent('Â£99.99');
      expect(screen.getByText('Â£129.99')).toBeInTheDocument();
    });
  });

  describe('sale tag', () => {
    it('should display the sale tag if the product is on sale', () => {
      setup(
        <ProductCard
          product={{ ...productData, isOnSale: true }}
          variant="horizontal"
        />
      );

      expect(screen.getByText(/sale/i)).toBeInTheDocument();
    });

    it('should not display the sale tag if the product is not on sale', () => {
      setup(
        <ProductCard
          product={{ ...productData, isOnSale: false }}
          variant="horizontal"
        />
      );

      expect(screen.queryByText(/sale/i)).not.toBeInTheDocument();
    });
  });
});
