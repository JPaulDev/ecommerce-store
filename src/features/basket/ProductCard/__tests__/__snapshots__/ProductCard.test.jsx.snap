// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard component should render a product card with with all the product information 1`] = `
.c13 {
  background-image: linear-gradient( to top, var(--color-green-2), var(--color-green-5) );
  border-radius: 6px;
  box-shadow: var(--box-shadow-md);
  color: white;
  cursor: pointer;
  font-size: var(--font-size-12);
  font-weight: var(--font-weight-bold);
  margin-top: 8px;
  padding: 12px 0;
  text-transform: uppercase;
  width: 180px;
}

.c13:disabled {
  background: var(--color-grey-10);
  cursor: initial;
  box-shadow: none;
  pointer-events: none;
}

.c13:not(:disabled):hover {
  background-image: linear-gradient( to top, var(--color-green-1), var(--color-green-4) );
  box-shadow: var(--box-shadow-hover);
}

.c13:not(:disabled):active {
  -webkit-transform: scale(0.96);
  -ms-transform: scale(0.96);
  transform: scale(0.96);
}

.c4 {
  color: var(--color-blue-4);
  font-size: var(--font-size-12);
  font-weight: var(--font-weight-bold);
  text-transform: uppercase;
}

.c7 {
  color: var(--color-grey-7);
  font-size: var(--font-size-16);
  -webkit-text-decoration: line-through;
  text-decoration: line-through;
}

.c8 {
  font-size: 0.9em;
  font-weight: var(--font-weight-semibold);
}

.c5 {
  color: var(--color-blue-4);
  font-size: var(--font-size-30);
}

.c6 {
  font-size: 0.7em;
}

.c9 {
  margin-top: 8px;
  position: relative;
}

.c10 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background-color: var(--color-grey-14);
  border: 1px solid var(--color-grey-10);
  border-radius: 4px;
  box-shadow: 0 2px 3px rgb(17 17 17 / 10%);
  font-size: var(--font-size-14);
  font-weight: var(--font-weight-medium);
  padding: 5px 25px 5px 10px;
  width: 55px;
}

.c10:not(:disabled) {
  cursor: pointer;
}

.c10:not(:disabled):hover {
  background-color: var(--color-grey-12);
}

.c11 {
  --arrow-color: var(--color-grey-4);
  --arrow-size: 5px;
  height: 100%;
  pointer-events: none;
  position: absolute;
  right: 0;
  width: 25px;
}

.c11::before,
.c11::after {
  border-left: var(--arrow-size) solid transparent;
  border-right: var(--arrow-size) solid transparent;
  content: '';
  left: 50%;
  position: absolute;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
}

.c11::before {
  border-bottom: var(--arrow-size) solid var(--arrow-color);
  top: 9px;
}

.c11::after {
  border-top: var(--arrow-size) solid var(--arrow-color);
  bottom: 9px;
}

.c12 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: var(--color-green-5);
  -webkit-column-gap: 5px;
  column-gap: 5px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: var(--font-size-11);
  font-weight: var(--font-weight-bold);
  margin-top: 8px;
  text-transform: uppercase;
}

.c0 {
  background-color: white;
  border: 1px solid var(--color-grey-12);
  border-radius: 10px;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 100%;
  padding: 20px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c0 img {
  height: auto;
  width: 150px;
}

.c1 {
  -webkit-line-clamp: 2;
  margin: 5px 0 10px;
  -webkit-box-orient: vertical;
  display: -webkit-box;
  overflow: hidden;
  text-align: center;
}

.c2 {
  -webkit-line-clamp: 3;
  color: var(--color-grey-4);
  font-size: var(--font-size-14);
  line-height: 1.5;
  margin-top: 10px;
  -webkit-box-orient: vertical;
  display: -webkit-box;
  overflow: hidden;
  text-align: center;
}

.c3 {
  border-bottom: 2px solid var(--color-grey-11);
  margin: 5px 0;
  padding-bottom: 10px;
  width: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

@media screen and (min-width:570px) {
  .c0 {
    min-height: 510px;
  }
}

@media screen and (max-width:1109.98px) {
  .c0:nth-child(4n) {
    display: none;
  }
}

@media screen and (min-width:570px) {
  .c1 {
    margin: 5px 0 auto;
  }
}

<li
  className="c0"
>
  <img
    alt=""
    height={200}
    quality={85}
    src=""
    width={200}
  />
  <p
    className="c1"
  >
    ASUS ROG Strix GeForce RTX 3080 V2 OC
  </p>
  <p
    className="c2"
  >
    1440MHz, 1905MHz Boost, 8704 CUDA Cores, 10GB GDDR6X
  </p>
  <div
    className="c3"
  >
    <span
      className="c4"
    >
      SKU: 
      GPU-3PAQ-ASU
    </span>
  </div>
  <span
    className="c5"
  >
    <span
      className="c6"
    >
      £
    </span>
    99
    <span
      className="c6"
    >
      .99
    </span>
  </span>
  <span
    className="c7"
  >
    <span
      className="c8"
    >
      Was: 
    </span>
    £129.99
  </span>
  <div
    className="c9"
  >
    <label
      hidden={true}
      htmlFor=":r0:"
    >
      Quantity
    </label>
    <select
      className="c10"
      disabled={false}
      id=":r0:"
      name="quantity"
      onChange={[Function]}
      value={1}
    >
      <option
        value={1}
      >
        1
      </option>
      <option
        value={2}
      >
        2
      </option>
      <option
        value={3}
      >
        3
      </option>
      <option
        value={4}
      >
        4
      </option>
      <option
        value={5}
      >
        5
      </option>
    </select>
    <span
      className="c11"
    />
  </div>
  <div
    className="c12"
    title="Item is in stock"
  >
    <svg
      enableBackground="new 0 0 24 24"
      fill="var(--color-grey-4)"
      height={18}
      viewBox="0 0 24 24"
      width={18}
      xmlns="http://www.w3.org/2000/svg"
    >
      <g>
        <rect
          fill="none"
          height="24"
          width="24"
        />
      </g>
      <g>
        <g>
          <path
            d="M5,5h2v3h10V5h2v5h2V5c0-1.1-0.9-2-2-2h-4.18C14.4,1.84,13.3,1,12,1S9.6,1.84,9.18,3H5C3.9,3,3,3.9,3,5v14 c0,1.1,0.9,2,2,2h6v-2H5V5z M12,3c0.55,0,1,0.45,1,1s-0.45,1-1,1s-1-0.45-1-1S11.45,3,12,3z"
          />
          <polygon
            points="21,11.5 15.51,17 12.5,14 11,15.5 15.51,20 22.5,13"
          />
        </g>
      </g>
    </svg>
    In Stock
  </div>
  <button
    className="c13"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Add To Basket
  </button>
</li>
`;
